shimmingtoolbox/shim/sequencer.py

```diff
@@ -42,838 +38,1063 @@
                           1 voxel(e.g. (50x50x1). Refer to :func:`shimmingtoolbox.shim.sequencer.extend_slice`/
                           :func:`shimmingtoolbox.shim.sequencer.update_affine_for_ap_slices`
-        method (str): Supported optimizer: 'least_squares', 'pseudo_inverse'. Note: refer to their specific
+            method (str): Supported optimizer: 'least_squares', 'pseudo_inverse'. Note: refer to their specific
                       implementation to know limits of the methods in: :mod:`shimmingtoolbox.optimizer`
-        opt_criteria (str): Criteria for the optimizer 'least_squares'. Supported: 'mse': mean squared error,
+            opt_criteria (str): Criteria for the optimizer 'least_squares'. Supported: 'mse': mean squared error,
                             'mae': mean absolute error, 'std': standard deviation.
-        mask_dilation_kernel (str): kernel used to dilate the mask. Allowed shapes are: 'sphere', 'cross', 'line'
+            mask_dilation_kernel (str): kernel used to dilate the mask. Allowed shapes are: 'sphere', 'cross', 'line'
                                     'cube'. See :func:`shimmingtoolbox.masking.mask_utils.dilate_binary_mask` for more
                                     details.
-        mask_dilation_kernel_size (int): Length of a side of the 3d kernel to dilate the mask. Must be odd. For example,
+            mask_dilation_kernel_size (int): Length of a side of the 3d kernel to dilate the mask. Must be odd. For example,
                                          a kernel of size 3 will dilate the mask by 1 pixel.
-        reg_factor (float): Regularization factor for the current when optimizing. A higher coefficient will
+            reg_factor (float): Regularization factor for the current when optimizing. A higher coefficient will
                             penalize higher current values while a lower factor will lower the effect of the
                             regularization. A negative value will favour high currents (not preferred). Only relevant
                             for 'least_squares' opt_method.
-        path_output (str): Path to the directory to output figures. Set logging level to debug to output debug
+            path_output (str): Path to the directory to output figures. Set logging level to debug to output debug
                            artefacts.
+        """
+        self.nii_fieldmap = nii_fieldmap
+        self.nii_anat = nii_anat
+        self.nii_mask_anat = nii_mask_anat
+        self.slices = slices
+        self.coils = coils
+        self.method = method
+        self.opt_criteria = opt_criteria
+        self.mask_dilation_kernel = mask_dilation_kernel
+        self.mask_dilation_kernel_size = mask_dilation_kernel_size
+        self.reg_factor = reg_factor
+        self.path_output = path_output
+        self.optimizer = None
+        self.bounds = None
+        self.extending = True
+
+    def shim_sequencer(self):
```

Comment:
I think we should move most of the error checking in the init and have the "shim" part be mostly just that, shimming. You could create a local method that includes most of this error checking. I would think that there is a lot of duplication between dynamic sequencer and the realtime one so most of this error checking could probably move to the base sequencer class.

Reply:
Also as mentioned above, rename to shim()

Reply:
Consider using @property so you can error check every time an attribute is changed. This would allow to be cleaner rather than have a "one function that checks everything" and if one attribute changes then we have to call the function that checks everything again.

Reply:
I separated the two parts of the sequencer. However, I do think that is good to keep in the real time sequencer, the calculation of the parameters for the real time shimming. (Because it's not just checking the dimensions of everything, but also a lot of mathematical operation etc.)
If needed, I can also separate it in a different function

Reply:
@Aurelienpujolm asked me for some advice. I have no read all the code in detail, but I think

a. `check_inputs` should be deleted. My instinct is that checking preconditions  be done _in each method_, near where they are used. Perhaps some precondition checking should happen in `__init__`, but it is a lot easier to avoid bugs if each function that assumes it has a 3d array checks that it has 3d array. It means more code but it's boilerplate code that's easy to write and easy to fix if it's wrong, and it avoid the ambiguity that users will have about "when/where should I call this `check_inputs()` function?". Also `check_inputs()` isn't just checking inputs, it is editing them too, which is misleading:

   https://github.com/shimming-toolbox/shimming-toolbox/blob/fe37b5d1a725a2deaab7697a275489e16928d267/shimmingtoolbox/shim/sequencer.py#L220-L230

    There should only be one line that says `self.nii_fieldmap = ` and it should be in `__init__`. All the transformations/normalizations needed should be done to that data before it is saved as a member variable.
b. But `property()` isn't a good idea either. Immutability is. Simply disallow people editing the shim parameters. If they want to change the shim parameters, force them to make a new shim object, which then runs through all the precondition checking in `__init__()` (or maybe in `_check_inputs()`).

Reply:
> I think we should move most of the error checking in the init and have the "shim" part be mostly just that, shimming.

I also think you can go further and stop using a class at all: combine `__init__()` and `shim()` into a single **function** (**not a method**) that does all the work in one step, and returns an object with the results (meaning your `.coefs`, `.slices`, etc). Move all the helper methods like `plot_current` and `eval` to their own top-level functions.

An _object_ is supposed to be a little miniature state machine: some data plus some methods that manipulate that data to keep the different pieces of it consistent. But state machines are notoriously hard to get right; it's so easy to add a little edit to some method that subtly invalidates other pieces of data, but only if that method gets called in a certain order. It's a nightmare to keep track of. Immutability is such a much stronger technique that objects are just unnecessary, in my opinion.

I read the addition of [`dataclass`](https://docs.python.org/3/library/dataclasses.html) as python recognizing that objects are a mistake, and structs + top level functions that know how to interpret them are better.

----

And while avoiding objects, also avoid inheritance. Your life is better without it. It is difficult and unnecessary and has wasted a lot of money in the last 20 years. ["Composition" is better](https://www.digitalocean.com/community/tutorials/composition-vs-inheritance), which simply means: write subroutines. That doesn't mean you can't have polymorphism -- you can design APIs and then write a series of functions that fit them, a set that all have the same input and output types or you can design [protocols](https://peps.python.org/pep-0544/) (what other languages call 'interfaces') -- but you can do it without confusing yourself and future maintainers with trying to use inheritance.